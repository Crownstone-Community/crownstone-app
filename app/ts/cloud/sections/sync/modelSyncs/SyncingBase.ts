export class SyncingBase {
  actions: any[];
  transferPromises : any[];
  globalCloudIdMap : syncIdMap;
  globalLocalIdMap : syncIdMap;

  constructor(actions : any[], transferPromises: any[], globalCloudIdMap: syncIdMap) {
    if (!globalCloudIdMap) {
      globalCloudIdMap = getSyncIdMap();
    }

    this.globalCloudIdMap = globalCloudIdMap;
    this.actions = actions;
    this.transferPromises = transferPromises;
  }

  _constructLocalIdMap() {
    this.globalLocalIdMap = getSyncIdMap();
    let globalKeys = Object.keys(this.globalCloudIdMap);
    globalKeys.forEach((key) => {
      this.globalLocalIdMap[key] = {};

      let cloudIds = Object.keys(this.globalCloudIdMap[key]);
      cloudIds.forEach((cloudId) => {
        let localId = this.globalCloudIdMap[key][cloudId];
        this.globalLocalIdMap[key][localId] = cloudId;
      })
    });
  }

  _getLocalData(store) {}

  download(cloudSphereId) {
    // should be implemented if the data should be pulled by the sync script
  }

  sync(...any) {
    throw new Error("SyncingBase: NOT OVERLOADED sync");
  }

  syncDown(...any) {
    throw new Error("SyncingBase: NOT OVERLOADED syncDown");
  }

  syncUp(...any) {
    throw new Error("SyncingBase: NOT OVERLOADED syncUp");
  }

  syncChildren(...any) {
    // should be implemented if the
  }


  /**
   * This should give a map of keys : cloudId, value: local Id
   *
   * Rough outline:
   *

   let cloudIdMap = {};
   let thingIds = Object.keys(thingsInState);
   thingIds.forEach((localThingId) => {
      let thing = thingsInState[localThingId];
      let cloudIdInThing = thing.config.cloudId; // <---------- location of cloudId in data model can be different!
      if (cloudIdInThing) {
        cloudIdMap[cloudIdInThing] = localThingId;
      }
    });

   return cloudIdMap;

   * @param thingsInState
   * @private
   */
  _getCloudIdMap(thingsInState : any) {
    throw new Error("SyncingBase: NOT OVERLOADED _getCloudIdMap");
  }




  /**
   * This should give an Id of a locally stored object that should match the remote object.
   *
   * Rough outline:
   *

   let thingIds = Object.keys(thingsInState);
   for (let i = 0; i < thingIds.length; i++) {
      let thing = thingsInState[thingIds[i]];
      if (<NUMBER OF CONDITIONS TO MATCH LOCAL DATA WITH CLOUD DATA>) {
        return thingIds[i];
      }
   }
   return null;

   * @param thingsInState
   * @param thing_in_cloud
   * @private
   */
  _searchForLocalMatch(thingsInState, thing_in_cloud) {
    throw new Error("SyncingBase: NOT OVERLOADED _searchForLocalMatch");
  }
}

export class SyncingSphereItemBase extends SyncingBase {
  localSphereId   : string;
  cloudSphereId   : string;

  constructor(
    actions : any[],
    transferPromises: any[],
    localSphereId: string,
    cloudSphereId: string,
    globalCloudIdMap: syncIdMap
  ) {
    super(actions, transferPromises, globalCloudIdMap);
    this.localSphereId = localSphereId;
    this.cloudSphereId = cloudSphereId;
  }
}


export function getSyncIdMap() : syncIdMap {
  return {
    abilities:  {},
    abilityProperties:  {},
    behaviours:  {},
    devices:     {},
    locations:   {},
    messages:    {},
    preferences: {},
    stones:      {},
    scenes:      {},
    sortedLists: {},
    spheres:     {},
    schedules:   {},
    hubs:        {},
    toons:       {},
    users:       {},
  }
}