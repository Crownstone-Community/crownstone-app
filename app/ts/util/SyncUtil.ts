
export const SyncUtil = {

  /**
   * This function will take the input reply object and create all fields (if required) in the tree structure
   * Finally, it can place the data provided in the child data field.
   * @param reply
   * @param treeArray
   * @param data
   */
  constructReply(reply: SyncRequestSphereData, treeArray: string[], data?) {
    let tree : any = reply;
    let lastTree : any;
    for (let item of treeArray) {
      if (tree[item] === undefined) {
        tree[item] = {};
      }
      lastTree = tree;
      tree = tree[item];
    }

    if (data) {
      lastTree[treeArray[treeArray.length - 1]].data = data;
    }
  },

}