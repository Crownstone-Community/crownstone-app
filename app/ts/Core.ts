import { eventBus } from "./util/EventBus";
import { NativeBus } from "./native/libInterface/NativeBus";


export const core : core = {
  eventBus: eventBus,
  nativeBus: NativeBus,
  permissionState: {
    location:  'unknown',
    bluetooth: 'unknown',
    bluetoothType: 'SCANNER',
  },
  bleState: {
    bleAvailable: true,
    bleBroadcastAvailable: true,
  },
  store: {
    getState: () : ReduxAppState => {

      // @ts-ignore
      return {}
    },
    dispatch: () => {

    },
    batchDispatch: () => {

    }
  },
};

