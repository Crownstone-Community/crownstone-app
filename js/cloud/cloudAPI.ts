'use strict';
import { Alert } from 'react-native';
import { Util } from '../util/Util';

import { appliances }        from './sections/appliances'
import { bootloader }        from './sections/bootloader'
import { cloudApiBase }      from './sections/cloudApiBase'
import { devices }           from './sections/devices'
import { firmware }          from './sections/firmware'
import { fingerprints }      from './sections/fingerprints'
import { installations }     from './sections/installations'
import { locations }         from './sections/locations'
import { messages }          from './sections/messages'
import { stones }            from './sections/stones'
import { spheres }           from './sections/spheres'
import { schedules }         from './sections/schedules'
import { sync }              from './sections/sync/sync'
import { syncEvents }        from './sections/sync/syncEvents'
import { syncUsersInSphere } from './sections/sync/syncUsersInSphere'
import { user }              from './sections/user'

import { toon }              from './sections/thirdParty/toon'

function combineSections() {
  let result : any = {};
  Util.mixin(result, cloudApiBase);

  // mixin all modules.
  Util.mixin(result, appliances);
  Util.mixin(result, bootloader);
  Util.mixin(result, devices);
  Util.mixin(result, firmware);
  Util.mixin(result, fingerprints);
  Util.mixin(result, installations);
  Util.mixin(result, locations);
  Util.mixin(result, messages);
  Util.mixin(result, schedules);
  Util.mixin(result, spheres);
  Util.mixin(result, stones);
  Util.mixin(result, sync);
  Util.mixin(result, syncEvents);
  Util.mixin(result, syncUsersInSphere);
  Util.mixin(result, user);

  result["thirdParty"] = {toon:{}};
  Util.mixin(result.thirdParty.toon, toon);

  return result;
}

/**
 * This adds all sections into the CLOUD
 */
export const CLOUD : any = combineSections();

