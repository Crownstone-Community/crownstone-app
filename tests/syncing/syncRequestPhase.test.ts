import { resetMocks } from "../__testUtil/mocks/suite.mock";
import { addSphere, addStone } from "../__testUtil/helpers/data.helper";
import { core } from "../../app/ts/Core";
import { SyncNext } from "../../app/ts/cloud/sections/newSync/SyncNext";
import { getGlobalIdMap } from "../../app/ts/cloud/sections/sync/modelSyncs/SyncingBase";
import { reply1 } from "./replyMocks/reply1";
import { CLOUD } from "../../app/ts/cloud/cloudAPI";
import { SyncUtil } from "../../app/ts/util/SyncUtil";

beforeEach(async () => {
  resetMocks()
})
beforeAll(async () => {})
afterEach(async () => { })
afterAll(async () => {})

test("Using SyncUtil", async () => {
  let myReply = {}
  SyncUtil.constructReply(myReply,['item', 'id'],{ myData:'awesome' });
  expect(myReply).toStrictEqual({item:{id:{data:{myData:'awesome'}}}})
  let myReply2 = {}
  SyncUtil.constructReply(myReply2,['item'],{ myData:'awesome' });
  expect(myReply2).toStrictEqual({item:{data:{myData:'awesome'}}})
})

// let handle = 'TestHandle';
// test("Using core in tests", async () => {
//   let cloudIdMap = getGlobalIdMap();
//   let actions = []
//   await SyncNext.processSyncResponse(reply1, actions, cloudIdMap)
//   expect(actions.length > 0).toBeTruthy()
//
//   core.store.batchDispatch(actions);
//
//   let state = core.store.getState();
//   for (let sphereId in state.spheres) {
//     let sphere = state.spheres[sphereId];
//     expect(Object.keys(sphere.users).length > 0).toBeTruthy()
//     for (let stoneId in sphere.stones) {
//       let stone = sphere.stones[stoneId];
//       for (let abilityType in stone.abilities) {
//         expect(stone.abilities[abilityType].syncedToCrownstone).toBeTruthy()
//       }
//     }
//   }
// })
//
// test("check mock", async () => {
//   CLOUD.forLocation(123).downloadLocationPicture(5).then((x) => { console.log(x) })
// })

// test ("Handle Sync Reply", async () => {
//   let actions = [];
//   let globalIdMap = getGlobalIdMap()
//   await SyncNext.processSyncResponse(reply2 as any, actions, globalIdMap)
//   console.log("globalIdMap", globalIdMap)
//   console.log("actions", actions)
// })