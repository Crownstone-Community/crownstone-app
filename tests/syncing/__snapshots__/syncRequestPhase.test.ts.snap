// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if fingerprints are syncing 1`] = `
{
  "sphereCloudId": {
    "data": {
      "gpsLocation": {
        "lat": null,
        "lng": null,
      },
      "name": "testSphere",
      "uid": null,
      "updatedAt": "2022-01-01T12:00:00.000Z",
      "uuid": "uuid2006-2007",
    },
    "fingerprints": {
      "uuid2035-2036": {
        "data": {
          "createdAt": "2022-01-01T12:00:00.000Z",
          "createdByUser": "user_uuid2002-2003",
          "createdOnDeviceType": "undefined_undefined",
          "crownstonesAtCreation": [
            "56541_57055",
            "58597_59111",
            "60653_61167",
          ],
          "data": [
            {
              "data": {
                "56541_57055": -65,
                "58597_59111": -65,
                "60653_61167": -65,
              },
              "dt": 0,
            },
          ],
          "locationId": "locationCloud",
          "type": "IN_HAND",
          "updatedAt": "2022-01-01T12:00:00.000Z",
        },
        "new": true,
      },
    },
    "hubs": {},
    "locations": {
      "locationCloud": {
        "data": {
          "updatedAt": 1641038400000,
        },
      },
    },
    "messages": {},
    "scenes": {},
    "stones": {
      "stone_uuid2008-2009": {
        "abilities": {
          "dimming": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "dimming",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "softOnSpeed": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "softOnSpeed",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 8,
                },
                "new": true,
              },
            },
          },
          "switchcraft": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "switchcraft",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "defaultDimValue": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "defaultDimValue",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 0,
                },
                "new": true,
              },
              "doubleTapSwitchcraft": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "doubleTapSwitchcraft",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": false,
                },
                "new": true,
              },
            },
          },
          "tapToToggle": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "tapToToggle",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "rssiOffset": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "rssiOffset",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 0,
                },
                "new": true,
              },
            },
          },
        },
        "behaviours": {},
        "data": {
          "address": undefined,
          "bootloaderVersion": null,
          "description": "",
          "firmwareVersion": "5.4.0",
          "hardwareVersion": null,
          "hidden": false,
          "icon": "c2-pluginFilled",
          "locationId": null,
          "locked": false,
          "major": 56541,
          "minor": 57055,
          "name": "stone_uuid2011",
          "type": "PLUG",
          "uicr": null,
          "uid": 1,
          "updatedAt": "2022-01-01T12:00:00.000Z",
        },
        "new": true,
      },
      "stone_uuid2016-2017": {
        "abilities": {
          "dimming": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "dimming",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "softOnSpeed": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "softOnSpeed",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 8,
                },
                "new": true,
              },
            },
          },
          "switchcraft": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "switchcraft",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "defaultDimValue": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "defaultDimValue",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 0,
                },
                "new": true,
              },
              "doubleTapSwitchcraft": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "doubleTapSwitchcraft",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": false,
                },
                "new": true,
              },
            },
          },
          "tapToToggle": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "tapToToggle",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "rssiOffset": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "rssiOffset",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 0,
                },
                "new": true,
              },
            },
          },
        },
        "behaviours": {},
        "data": {
          "address": undefined,
          "bootloaderVersion": null,
          "description": "",
          "firmwareVersion": "5.4.0",
          "hardwareVersion": null,
          "hidden": false,
          "icon": "c2-pluginFilled",
          "locationId": null,
          "locked": false,
          "major": 58597,
          "minor": 59111,
          "name": "stone_uuid2019",
          "type": "PLUG",
          "uicr": null,
          "uid": 2,
          "updatedAt": "2022-01-01T12:00:00.000Z",
        },
        "new": true,
      },
      "stone_uuid2024-2025": {
        "abilities": {
          "dimming": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "dimming",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "softOnSpeed": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "softOnSpeed",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 8,
                },
                "new": true,
              },
            },
          },
          "switchcraft": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "switchcraft",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "defaultDimValue": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "defaultDimValue",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 0,
                },
                "new": true,
              },
              "doubleTapSwitchcraft": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "doubleTapSwitchcraft",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": false,
                },
                "new": true,
              },
            },
          },
          "tapToToggle": {
            "data": {
              "enabled": false,
              "syncedToCrownstone": true,
              "type": "tapToToggle",
              "updatedAt": "2022-01-01T12:00:00.000Z",
            },
            "new": true,
            "properties": {
              "rssiOffset": {
                "data": {
                  "syncedToCrownstone": true,
                  "type": "rssiOffset",
                  "updatedAt": "2022-01-01T12:00:00.000Z",
                  "value": 0,
                },
                "new": true,
              },
            },
          },
        },
        "behaviours": {},
        "data": {
          "address": undefined,
          "bootloaderVersion": null,
          "description": "",
          "firmwareVersion": "5.4.0",
          "hardwareVersion": null,
          "hidden": false,
          "icon": "c2-pluginFilled",
          "locationId": null,
          "locked": false,
          "major": 60653,
          "minor": 61167,
          "name": "stone_uuid2027",
          "type": "PLUG",
          "uicr": null,
          "uid": 3,
          "updatedAt": "2022-01-01T12:00:00.000Z",
        },
        "new": true,
      },
    },
    "toons": {},
    "users": {},
  },
}
`;

exports[`Check if fingerprints are syncing 2`] = `
{
  "app": {
    "activeSphere": null,
    "hasSeenDeviceSettings": false,
    "hasSeenSwitchView": false,
    "hasZoomedOutForSphereOverview": false,
    "indoorLocalizationEnabled": true,
    "migratedDataToVersion": null,
    "notificationToken": null,
    "tapToToggleEnabled": false,
    "updatedAt": 1,
  },
  "devApp": {
    "fastPhone": false,
    "sphereUsedForSetup": "DEV_APP_SPHERE_ID",
    "storeCrownstonesInCloud": false,
  },
  "development": {
    "debugRTCtoken": null,
    "devAppVisible": false,
    "firmwareEarlyAccessLevel": 0,
    "log_advertisements": 50,
    "log_ble": 50,
    "log_cloud": 30,
    "log_constellation": 30,
    "log_dfu": 30,
    "log_events": 50,
    "log_info": 30,
    "log_native": 50,
    "log_nav": 30,
    "log_notifications": 30,
    "log_scheduler": 50,
    "log_store": 30,
    "logging_enabled": false,
    "nativeExtendedLogging": false,
    "preview": false,
    "show_rssi_values_in_mesh": false,
    "show_sync_button_in_behaviour": false,
    "useHighFrequencyScanningInBackground": false,
    "use_time_based_nonce": false,
  },
  "devices": {},
  "events": {
    "remove": {
      "behaviours": {},
      "devices": {},
      "fingerprints": {},
      "installations": {},
      "locations": {},
      "messages": {},
      "scenes": {},
      "stones": {},
      "user": {},
    },
    "special": {
      "behaviours": {},
      "devices": {},
      "fingerprints": {},
      "installations": {},
      "locations": {},
      "messages": {},
      "scenes": {},
      "stones": {},
      "user": {},
    },
  },
  "installations": {},
  "spheres": {
    "sphere_uuid2004-2005": {
      "config": {
        "cloudId": "sphereCloudId",
        "iBeaconUUID": "uuid2006-2007",
        "latitude": null,
        "longitude": null,
        "name": "testSphere",
        "uid": null,
        "updatedAt": 1641038400000,
      },
      "hubs": {},
      "id": "sphere_uuid2004-2005",
      "keys": {},
      "locations": {
        "location_uuid2032-2033": {
          "config": {
            "cloudId": "locationCloud",
            "fingerprintCloudId": null,
            "fingerprintParsed": null,
            "fingerprintRaw": null,
            "fingerprintUpdatedAt": 1,
            "icon": undefined,
            "name": "stone_uuid2034",
            "picture": null,
            "pictureId": null,
            "pictureSource": null,
            "pictureTaken": null,
            "uid": null,
            "updatedAt": 1641038400000,
          },
          "fingerprints": {
            "processed": {},
            "raw": {
              "uuid2035-2036": {
                "cloudId": null,
                "createdAt": 1641038400000,
                "createdByUser": "user_uuid2002-2003",
                "createdOnDeviceType": "undefined_undefined",
                "crownstonesAtCreation": {
                  "56541_57055": true,
                  "58597_59111": true,
                  "60653_61167": true,
                },
                "data": [
                  {
                    "data": {
                      "56541_57055": -65,
                      "58597_59111": -65,
                      "60653_61167": -65,
                    },
                    "dt": 0,
                  },
                ],
                "id": "uuid2035-2036",
                "type": "IN_HAND",
                "updatedAt": 1641038400000,
              },
              "uuid2037-2038": {
                "cloudId": "newCloudFingerprint",
                "createdAt": 1609459200000,
                "createdByUser": "otherUser",
                "createdOnDeviceType": "test_test",
                "crownstonesAtCreation": {
                  "10_10": true,
                },
                "data": [
                  {
                    "data": {
                      "10_10": -52,
                    },
                    "dt": 123,
                  },
                ],
                "id": "uuid2037-2038",
                "type": "IN_POCKET",
                "updatedAt": 1640995200000,
              },
            },
          },
          "id": "location_uuid2032-2033",
          "layout": {
            "x": null,
            "y": null,
          },
          "presentUsers": [],
        },
      },
      "messages": {},
      "scenes": {},
      "sortedLists": {},
      "state": {
        "lastPresentTime": 0,
        "present": false,
        "reachable": false,
        "smartHomeEnabled": true,
      },
      "stones": {
        "stone_uuid2008-2009": {
          "abilities": {
            "dimming": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "softOnSpeed": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "softOnSpeed",
                  "updatedAt": 1641038400000,
                  "value": 8,
                  "valueTarget": 8,
                },
              },
              "syncedToCrownstone": true,
              "type": "dimming",
              "updatedAt": 1641038400000,
            },
            "switchcraft": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "defaultDimValue": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "defaultDimValue",
                  "updatedAt": 1641038400000,
                  "value": 0,
                  "valueTarget": 0,
                },
                "doubleTapSwitchcraft": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "doubleTapSwitchcraft",
                  "updatedAt": 1641038400000,
                  "value": false,
                  "valueTarget": false,
                },
              },
              "syncedToCrownstone": true,
              "type": "switchcraft",
              "updatedAt": 1641038400000,
            },
            "tapToToggle": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "rssiOffset": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "rssiOffset",
                  "updatedAt": 1641038400000,
                  "value": 0,
                  "valueTarget": 0,
                },
              },
              "syncedToCrownstone": true,
              "type": "tapToToggle",
              "updatedAt": 1641038400000,
            },
          },
          "behaviours": {},
          "config": {
            "bootloaderVersion": null,
            "cloudId": null,
            "description": "",
            "dfuResetRequired": false,
            "firmwareVersion": "5.4.0",
            "handle": "handle1",
            "hardwareVersion": null,
            "hidden": false,
            "iBeaconMajor": 56541,
            "iBeaconMinor": 57055,
            "icon": "c2-pluginFilled",
            "locationId": null,
            "locked": false,
            "macAddress": undefined,
            "name": "stone_uuid2011",
            "type": "PLUG",
            "uicr": null,
            "uid": 1,
            "updatedAt": 1641038400000,
          },
          "errors": {
            "dimmerOffFailure": false,
            "dimmerOnFailure": false,
            "hasError": false,
            "overCurrent": false,
            "overCurrentDimmer": false,
            "temperatureChip": false,
            "temperatureDimmer": false,
          },
          "id": "stone_uuid2008-2009",
          "keys": {},
          "lastUpdated": {
            "stoneTime": 0,
          },
          "reachability": {
            "lastSeen": null,
          },
          "state": {
            "behaviourOverridden": false,
            "currentUsage": 0,
            "dimmerReady": false,
            "powerFactor": null,
            "previousState": 0,
            "state": 0,
            "timeSet": false,
            "updatedAt": 1,
          },
        },
        "stone_uuid2016-2017": {
          "abilities": {
            "dimming": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "softOnSpeed": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "softOnSpeed",
                  "updatedAt": 1641038400000,
                  "value": 8,
                  "valueTarget": 8,
                },
              },
              "syncedToCrownstone": true,
              "type": "dimming",
              "updatedAt": 1641038400000,
            },
            "switchcraft": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "defaultDimValue": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "defaultDimValue",
                  "updatedAt": 1641038400000,
                  "value": 0,
                  "valueTarget": 0,
                },
                "doubleTapSwitchcraft": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "doubleTapSwitchcraft",
                  "updatedAt": 1641038400000,
                  "value": false,
                  "valueTarget": false,
                },
              },
              "syncedToCrownstone": true,
              "type": "switchcraft",
              "updatedAt": 1641038400000,
            },
            "tapToToggle": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "rssiOffset": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "rssiOffset",
                  "updatedAt": 1641038400000,
                  "value": 0,
                  "valueTarget": 0,
                },
              },
              "syncedToCrownstone": true,
              "type": "tapToToggle",
              "updatedAt": 1641038400000,
            },
          },
          "behaviours": {},
          "config": {
            "bootloaderVersion": null,
            "cloudId": null,
            "description": "",
            "dfuResetRequired": false,
            "firmwareVersion": "5.4.0",
            "handle": "handle2",
            "hardwareVersion": null,
            "hidden": false,
            "iBeaconMajor": 58597,
            "iBeaconMinor": 59111,
            "icon": "c2-pluginFilled",
            "locationId": null,
            "locked": false,
            "macAddress": undefined,
            "name": "stone_uuid2019",
            "type": "PLUG",
            "uicr": null,
            "uid": 2,
            "updatedAt": 1641038400000,
          },
          "errors": {
            "dimmerOffFailure": false,
            "dimmerOnFailure": false,
            "hasError": false,
            "overCurrent": false,
            "overCurrentDimmer": false,
            "temperatureChip": false,
            "temperatureDimmer": false,
          },
          "id": "stone_uuid2016-2017",
          "keys": {},
          "lastUpdated": {
            "stoneTime": 0,
          },
          "reachability": {
            "lastSeen": null,
          },
          "state": {
            "behaviourOverridden": false,
            "currentUsage": 0,
            "dimmerReady": false,
            "powerFactor": null,
            "previousState": 0,
            "state": 0,
            "timeSet": false,
            "updatedAt": 1,
          },
        },
        "stone_uuid2024-2025": {
          "abilities": {
            "dimming": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "softOnSpeed": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "softOnSpeed",
                  "updatedAt": 1641038400000,
                  "value": 8,
                  "valueTarget": 8,
                },
              },
              "syncedToCrownstone": true,
              "type": "dimming",
              "updatedAt": 1641038400000,
            },
            "switchcraft": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "defaultDimValue": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "defaultDimValue",
                  "updatedAt": 1641038400000,
                  "value": 0,
                  "valueTarget": 0,
                },
                "doubleTapSwitchcraft": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "doubleTapSwitchcraft",
                  "updatedAt": 1641038400000,
                  "value": false,
                  "valueTarget": false,
                },
              },
              "syncedToCrownstone": true,
              "type": "switchcraft",
              "updatedAt": 1641038400000,
            },
            "tapToToggle": {
              "cloudId": null,
              "enabled": false,
              "enabledTarget": false,
              "properties": {
                "rssiOffset": {
                  "cloudId": null,
                  "syncedToCrownstone": true,
                  "type": "rssiOffset",
                  "updatedAt": 1641038400000,
                  "value": 0,
                  "valueTarget": 0,
                },
              },
              "syncedToCrownstone": true,
              "type": "tapToToggle",
              "updatedAt": 1641038400000,
            },
          },
          "behaviours": {},
          "config": {
            "bootloaderVersion": null,
            "cloudId": null,
            "description": "",
            "dfuResetRequired": false,
            "firmwareVersion": "5.4.0",
            "handle": "handle3",
            "hardwareVersion": null,
            "hidden": false,
            "iBeaconMajor": 60653,
            "iBeaconMinor": 61167,
            "icon": "c2-pluginFilled",
            "locationId": null,
            "locked": false,
            "macAddress": undefined,
            "name": "stone_uuid2027",
            "type": "PLUG",
            "uicr": null,
            "uid": 3,
            "updatedAt": 1641038400000,
          },
          "errors": {
            "dimmerOffFailure": false,
            "dimmerOnFailure": false,
            "hasError": false,
            "overCurrent": false,
            "overCurrentDimmer": false,
            "temperatureChip": false,
            "temperatureDimmer": false,
          },
          "id": "stone_uuid2024-2025",
          "keys": {},
          "lastUpdated": {
            "stoneTime": 0,
          },
          "reachability": {
            "lastSeen": null,
          },
          "state": {
            "behaviourOverridden": false,
            "currentUsage": 0,
            "dimmerReady": false,
            "powerFactor": null,
            "previousState": 0,
            "state": 0,
            "timeSet": false,
            "updatedAt": 1,
          },
        },
      },
      "thirdParty": {
        "toons": {},
      },
      "users": {},
    },
  },
  "user": {
    "accessToken": null,
    "appIdentifier": null,
    "betaAccess": false,
    "bootloaderVersionsAvailable": {},
    "developer": false,
    "email": null,
    "firmwareVersionsAvailable": {},
    "firstName": "Test",
    "isNew": true,
    "language": null,
    "lastName": "LastName",
    "passwordHash": null,
    "picture": null,
    "pictureId": null,
    "seenTapToToggle": false,
    "seenTapToToggleDisabledDuringSetup": false,
    "updatedAt": 1,
    "uploadDeviceDetails": true,
    "uploadLocation": true,
    "uploadSwitchState": true,
    "userId": "user_uuid2002-2003",
  },
}
`;

exports[`Check if messages are preparing their sync request payload 1`] = `
{
  "sphereCloudId": {
    "messages": {
      "cloudMessage": {
        "data": {
          "recipients": [
            "memberId",
          ],
          "updatedAt": 1641038400000,
        },
        "deletedBy": {},
        "readBy": {},
      },
      "messageId_uuid2008-2009": {
        "data": {
          "content": "testMessage",
          "everyoneInSphere": false,
          "includeSenderInEveryone": false,
          "recipients": [
            "memberId",
          ],
          "triggerEvent": "enter",
          "updatedAt": "2022-01-01T12:00:00.000Z",
        },
        "deletedBy": {
          "deleted": {
            "data": {
              "updatedAt": "2022-01-01T12:00:00.000Z",
              "userId": "user_uuid2002-2003",
            },
            "new": true,
          },
        },
        "new": true,
        "readBy": {
          "read": {
            "data": {
              "updatedAt": "2022-01-01T12:00:00.000Z",
              "userId": "user_uuid2002-2003",
            },
            "new": true,
          },
        },
      },
    },
  },
}
`;

exports[`Check if messages can handle the cloud's sync reply: creationActions 1`] = `
[
  {
    "data": {
      "cloudId": "dbId:MessageV2Repository:1",
      "content": "hello",
      "everyoneInSphere": false,
      "includeSenderInEveryone": false,
      "recipients": {
        "dbId:UserRepository:3": true,
      },
      "senderId": undefined,
      "sentAt": 0,
      "triggerEvent": "enter",
      "triggerLocationId": undefined,
      "updatedAt": 0,
    },
    "messageId": "uuid2008-2009",
    "sphereId": "sphere_uuid2004-2005",
    "type": "ADD_MESSAGE",
  },
]
`;
